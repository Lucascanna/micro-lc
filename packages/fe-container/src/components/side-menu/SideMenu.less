@import "../../styles/_variables";
@import "../../styles/_colors";

@side-menu-width: 256px;

.sideMenu {
  height: 100%;
  width: 256px;
  position: fixed;
  z-index: 999;
  top: @top-bar-height;
  left: -@side-menu-width;
  background-color: white;
  overflow-x: hidden;
  overflow-y: auto;
  transition: left .2s linear;
  padding-top: 24px;
  box-shadow: none;

  &.opened {
    left: 0;
    box-shadow: 0 8px 10px -5px rgba(0, 0, 0, 0.15), 0 16px 24px 2px rgba(0, 0, 0, 0.09), 0 6px 30px 5px rgba(0, 0, 0, 0.07);
  }
}

.sideMenu_overlay {
  display: none;
  position: fixed;
  top: @top-bar-height;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 998;
  background-color: @overlay-background;

  &.sideMenu_visible {
    display: block;
  }
}

.sideMenu_voice {
  cursor: pointer;
  display: grid;
  grid-template-columns: min-content 1fr;
  column-gap: 12px;
  align-items: center;
  padding-left: 24px;

  .sideMenu_icon {
    font-size: 16px;
    color: var(--primary-color);
  }

  .sideMenu_entry {
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: space-between;
    padding: 11px 11px 11px 0;

    .sideMenu_label {
      color: @menu-label-color;
      min-width: 160px;
    }

    .sideMenu_externalLink {
      color: var(--primary-color);
      margin-left: 4px;
    }
  }

  &:hover .sideMenu_label {
    color: var(--primary-color);
  }

  &.active {
    background: @menu-hover-bg-color;
    box-shadow: inset 3px 0 0 0 var(--primary-color);

    .sideMenu_label {
      color: var(--primary-color);
    }
  }
}
