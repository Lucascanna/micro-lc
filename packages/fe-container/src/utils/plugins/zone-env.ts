// eslint-disable-file
// @ts-nocheck
if (typeof global !== 'undefined') {
  global.__Zone_disable_registerElement = true
  global.__Zone_disable_customElements = true
}

export {}
