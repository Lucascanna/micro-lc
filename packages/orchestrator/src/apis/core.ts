import type { ImportMap } from '@micro-lc/interfaces'

import type { CompleteConfig } from '../config'

import type { BaseExtension } from './extensions'
import type MicroLC from './micro-lc'

export interface MicrolcApi<T extends BaseExtension> {
  readonly applyImportMap: (id: string, importmap: ImportMap) => void
  readonly extensions: Readonly<T>
  readonly getApplications: () => Readonly<CompleteConfig['applications']>
  readonly getCurrentConfig: () => Readonly<CompleteConfig>
  readonly router: {
    goToApplication<S = unknown>(id: string, opts?: {data?: S; type?: 'push' | 'replace'}): void
    open: (url: string | URL | undefined, target?: string | undefined, features?: string | undefined) => void
  }
  readonly setCurrentConfig: (newConfig: CompleteConfig) => void
  readonly setExtension: (key: keyof T, value: T[keyof T]) => Readonly<T>
  readonly shared: Readonly<Record<string, unknown>>
}

export function createMicrolcApiInstance<Extensions extends BaseExtension>(
  this: MicroLC<Extensions>
): () => MicrolcApi<Extensions> {
  const getApi = () => Object.freeze({
    applyImportMap: (id: string, importmap: ImportMap) => { this.applicationsImportMaps.set(id, importmap) },
    extensions: Object.freeze({ ...this.extensions }),
    getApplications: () => Object.freeze([...this.config.applications]),
    getCurrentConfig: () => Object.freeze({ ...this.config }),
    router: {
      goToApplication: (_id: string, data?: unknown): void => {
        const url = this.config.applications.find(({ id }) => id === _id)?.route
        if (url) {
          window.history.pushState(data, '', url)
        }
      },
      open: (url: string | URL | undefined, target?: string | undefined, features?: string | undefined) => {
        window.open(url, target, features)
      },
    },
    setCurrentConfig: (newConfig: CompleteConfig) => { this.config = newConfig },
    setExtension: (key: keyof Extensions, value: Extensions[keyof Extensions]) => {
      this.extensions[key] = value
      return Object.freeze({ ...this.extensions })
    },
    shared: Object.freeze({ ...this.config.shared }),
  })

  if (process.env.NODE_ENV === 'development') {
    Object.defineProperty(window, 'getMicrolcApi', { value: getApi })
  }

  return getApi
}
