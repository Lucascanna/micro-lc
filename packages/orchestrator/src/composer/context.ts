import type { Content } from '@micro-lc/interfaces'
import type { RenderOptions } from 'lit-html'
import { html, render } from 'lit-html'

import { interpolate } from './compiler'
import { jsonToHtml } from './json'
import { lexer } from './lexer'

interface ComposerOptions {
  context?: Record<string, unknown>
  extraProperties?: string[]
}

type ComposerContextAppender =
  (container: HTMLElement | DocumentFragment, options?: RenderOptions)
    => void

export type ExtendedHTMLElement<S extends Record<string, unknown>, T extends HTMLElement = HTMLElement> =
  T & S

export async function createComposerContext(
  layout: Content,
  { extraProperties, context = {} }: ComposerOptions = {}
): Promise<ComposerContextAppender> {
  const htmlBuffer = jsonToHtml(layout, extraProperties)
  const template = await lexer(htmlBuffer)
  const variables = interpolate(template.variables, context)
  return (container, options) => {
    render(html(template.literals, ...variables), container, options)
  }
}
