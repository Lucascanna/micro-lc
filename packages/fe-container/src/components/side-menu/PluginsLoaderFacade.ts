import {HrefConfig} from './SideMenu'

export interface PluginStrategy {
  handlePluginLoad: () => void
}

interface StrategyParam {
  pluginStrategy: string,
  hrefConfig: {
    sameWindow: boolean,
    url: string
  }
}

export const retrievePluginLoadingStrategy = (menuEntry: StrategyParam) => {
  switch (menuEntry.pluginStrategy) {
    case 'href':
    default:
      return hrefStrategy(menuEntry.hrefConfig)
  }
}

function hrefStrategy (hrefConfig?: HrefConfig | null): PluginStrategy {
  return {
    handlePluginLoad: () => {
      if (hrefConfig?.sameWindow) {
        window.location.href = hrefConfig.url
      } else if (hrefConfig?.url) {
        window.open(hrefConfig.url)
      }
    }
  }
}
