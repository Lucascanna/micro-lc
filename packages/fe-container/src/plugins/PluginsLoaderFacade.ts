import {createBrowserHistory} from 'history'
import {registerMicroApps, RegistrableApp, start} from 'qiankun'
import {ExternalLink, Plugin, User} from '@mia-platform/core'

import {INTEGRATION_METHODS} from '@constants'

export interface PluginStrategy {
  handlePluginLoad: () => void
}

const registeredPluginsStrategies = new Map<string, PluginStrategy>()
const registeredPlugins: Plugin[] = []

export const registerPlugin = (plugin: Plugin) => {
  const pluginStrategy: PluginStrategy = strategyBuilder(plugin)
  registeredPlugins.push(plugin)
  registeredPluginsStrategies.set(plugin.id, pluginStrategy)
}

export const retrievePluginStrategy = (plugin: Plugin) => {
  return registeredPluginsStrategies.get(plugin.id) || noOpStrategy()
}

export const isPluginLoaded = (plugin: Plugin) =>
  plugin.pluginRoute ? window.location.pathname.endsWith(plugin.pluginRoute) : false

export const isCurrentPluginLoaded = () => {
  return registeredPlugins.findIndex(isPluginLoaded) !== -1
}

const strategyBuilder = (plugin: Plugin) => {
  switch (plugin.integrationMode) {
    case INTEGRATION_METHODS.HREF:
      return hrefStrategy(plugin.externalLink)
    case INTEGRATION_METHODS.QIANKUN:
    case INTEGRATION_METHODS.IFRAME:
      return routeStrategy(plugin)
    default:
      return noOpStrategy()
  }
}

export const finish = (user: Partial<User>) => {
  const quiankunConfig = registeredPlugins
    .filter(plugin => plugin.integrationMode === INTEGRATION_METHODS.QIANKUN)
    .map<RegistrableApp<any>>(plugin => ({
      name: plugin.id,
      entry: plugin.pluginUrl || '',
      container: `#${plugin.id}`,
      activeRule: plugin.pluginRoute || '',
      props: {
        currentUser: user
      }
    }))
  registerMicroApps(quiankunConfig)
  start()
}

export const history = createBrowserHistory({basename: process.env.PUBLIC_URL})

function hrefStrategy (externalLink?: ExternalLink): PluginStrategy {
  return {
    handlePluginLoad: () => {
      if (externalLink?.sameWindow) {
        window.location.href = externalLink.url
      } else if (externalLink?.url) {
        window.open(externalLink.url)
      }
    }
  }
}

function routeStrategy (plugin: Plugin): PluginStrategy {
  return {
    handlePluginLoad: () => {
      history.push(plugin?.pluginRoute || '')
    }
  }
}

function noOpStrategy (): PluginStrategy {
  return {
    handlePluginLoad: () => {
    }
  }
}
