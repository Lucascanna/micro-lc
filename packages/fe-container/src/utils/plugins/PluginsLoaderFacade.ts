/*
 * Copyright 2021 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import {createBrowserHistory} from 'history'
import {addErrorHandler, registerMicroApps, RegistrableApp, start} from 'qiankun'
import {Plugin, User} from '@mia-platform/core'

import {ERROR_PATH, INTEGRATION_METHODS, MICROLC_QIANKUN_CONTAINER} from '@constants'
import {noOpStrategy} from '@utils/plugins/strategies/NoOpStrategy'
import {hrefStrategy} from '@utils/plugins/strategies/HrefStrategy'
import {routeStrategy} from '@utils/plugins/strategies/RouteStrategy'
import {PluginStrategy} from '@utils/plugins/strategies/PluginStrategy'

const registeredPluginsStrategies = new Map<string, PluginStrategy>()
const registeredPlugins: Plugin[] = []

export const registerPlugin = (plugin: Plugin) => {
  const pluginStrategy: PluginStrategy = strategyBuilder(plugin)
  registeredPlugins.push(plugin)
  registeredPluginsStrategies.set(plugin.id, pluginStrategy)
}

export const retrievePluginStrategy = (plugin: Plugin) => {
  return registeredPluginsStrategies.get(plugin.id) || noOpStrategy()
}

export const isPluginLoaded = (plugin: Plugin) =>
  plugin.pluginRoute ? window.location.pathname.endsWith(plugin.pluginRoute) : false

export const findCurrentPlugin = () => {
  return registeredPlugins.find(isPluginLoaded)
}

export const isCurrentPluginLoaded = () => {
  const isErrorPage = window.location.pathname.endsWith(ERROR_PATH.INTERNAL_ERROR) || window.location.pathname.endsWith(ERROR_PATH.UNAUTHORIZED)
  return isErrorPage || findCurrentPlugin() !== undefined
}

const strategyBuilder = (plugin: Plugin) => {
  switch (plugin.integrationMode) {
    case INTEGRATION_METHODS.HREF:
      return hrefStrategy(plugin.externalLink)
    case INTEGRATION_METHODS.QIANKUN:
    case INTEGRATION_METHODS.IFRAME:
      return routeStrategy(plugin)
    default:
      return noOpStrategy()
  }
}

export const finish = (user: Partial<User>) => {
  const basePath = retrieveBasePath()
  history = createBrowserHistory({basename: basePath})
  const quiankunConfig = registeredPlugins
    .filter(plugin => plugin.integrationMode === INTEGRATION_METHODS.QIANKUN)
    .map<RegistrableApp<any>>(plugin => ({
      name: plugin.id,
      entry: plugin.pluginUrl || '',
      container: `#${MICROLC_QIANKUN_CONTAINER}`,
      activeRule: `${basePath}${plugin.pluginRoute || ''}`,
      props: {
        basePath,
        currentUser: user
      }
    }))
  registerMicroApps(quiankunConfig)
  addErrorHandler(_ => history.push(ERROR_PATH.INTERNAL_ERROR))
  start()
}

const DOUBLE_SLASH = /\/\//g

const retrieveBasePath = () => {
  let basePath = `${window.location.pathname || ''}`
  const currentPlugin = findCurrentPlugin()
  if (currentPlugin?.pluginRoute) {
    basePath = window.location.pathname.replace(currentPlugin.pluginRoute, '')
  }
  basePath = basePath.replace(new RegExp(ERROR_PATH.INTERNAL_ERROR, 'g'), '')
    .replace(new RegExp(ERROR_PATH.UNAUTHORIZED, 'g'), '')
    .replace(DOUBLE_SLASH, '/')
  return basePath.endsWith('/') ? basePath.slice(0, -1) : basePath
}

export let history = createBrowserHistory({basename: window.location.pathname})
