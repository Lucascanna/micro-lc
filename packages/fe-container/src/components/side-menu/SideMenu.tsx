import React, {useCallback, useContext, useState} from 'react'
import PropTypes from 'prop-types'
import {Plugin} from '@mia-platform/core'

import {MenuOpenedContext} from '../../contexts/MenuOpened.context'
import {isPluginLoaded, PluginStrategy, retrievePluginStrategy} from '../../plugins/PluginsLoaderFacade'

import './SideMenu.less'

const sideMenuProps = {
  plugins: PropTypes.array
}

type SideMenuProps = PropTypes.InferProps<typeof sideMenuProps>

export const SideMenu: React.FC<SideMenuProps> = ({plugins}) => {
  const {isMenuOpened, setMenuOpened} = useContext(MenuOpenedContext)

  const closeMenu = useCallback(() => setMenuOpened(false), [setMenuOpened])

  const entriesMapper = useCallback((plugin: Plugin) => <SideMenuEntry key={plugin.id} {...plugin}/>, [])

  return (
    <>
      <div
        className={'sideMenu_overlay ' + (isMenuOpened ? '' : 'closed')} data-testid="layout-content-overlay"
        onClick={closeMenu}
      />
      <div className={'sideMenu ' + (isMenuOpened ? 'opened' : '')}>
        {plugins?.map(entriesMapper)}
      </div>
    </>
  )
}

SideMenu.propTypes = sideMenuProps

const SideMenuEntry: React.FC<Plugin> = (plugin) => {
  const [isActive, setIsActive] = useState<boolean>(isPluginLoaded(plugin))

  const menuClick = useCallback((plugin: Plugin) => {
    const pluginStrategy: PluginStrategy = retrievePluginStrategy(plugin)
    return () => {
      pluginStrategy.handlePluginLoad()
      setIsActive(isPluginLoaded(plugin))
    }
  }, [])

  return (
    <div className="sideMenu_voice" onClick={menuClick(plugin)}>
      <div className={'sideMenu_entry ' + (isActive ? 'active' : '')}>
        <i className={'sideMenu_icon ' + (plugin.icon || '')}/>
        <span className="sideMenu_label">{plugin.label}</span>
        {plugin.integrationMode === 'href' && <i className="fas fa-external-link-alt sideMenu_externalLink"/>}
      </div>
      <hr className="sideMenu_divider"/>
    </div>
  )
}
