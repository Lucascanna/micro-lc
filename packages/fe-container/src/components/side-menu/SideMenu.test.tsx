/*
 * Copyright 2021 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react'
import {screen} from '@testing-library/react'
import {Plugin} from '@mia-platform/core'
import userEvent from '@testing-library/user-event'

import {SideMenu} from './SideMenu'
import RenderWithReactIntl from '../../__tests__/utils'
import {registerPlugin} from '@utils/plugins/PluginsLoaderFacade'
import {MenuOpenedProvider} from '@contexts/MenuOpened.context'

describe('SideMenu tests', () => {
  afterAll(() => {
    jest.clearAllMocks()
    jest.resetAllMocks()
  })

  it('side menu show entries', () => {
    const entriesHref = {integrationMode: 'href', externalLink: {sameWindow: false, url: ''}}
    RenderWithReactIntl(
      <SideMenu plugins={[{label: 'entry_1', id: '1', ...entriesHref}, {label: 'entry_2', id: '2', ...entriesHref}]}/>
    )
    expect(screen.queryByText('entry_1')).toBeVisible()
    expect(screen.queryByText('entry_2')).toBeVisible()
    expect(screen.queryByText('entry_3')).toBeNull()
  })

  it('side menu click correctly works', async () => {
    window.open = jest.fn()
    const plugin: Plugin = {
      label: 'entry_1',
      id: '1',
      integrationMode: 'href',
      externalLink: {sameWindow: false, url: 'https://google.it'}
    }
    registerPlugin(plugin)
    RenderWithReactIntl(<SideMenu plugins={[plugin]}/>)
    userEvent.click(await screen.findByText('entry_1'))
    expect(window.open).toBeCalledWith('https://google.it')
  })

  it('display the icon', () => {
    window.open = jest.fn()
    RenderWithReactIntl(
      <SideMenu plugins={[{label: 'entry_1', id: '1', integrationMode: 'iframe', icon: 'test-icon'}]}/>
    )
    expect(document.getElementsByClassName('test-icon')).toHaveLength(1)
  })

  it('no icon, just placeholder class', () => {
    window.open = jest.fn()
    RenderWithReactIntl(<SideMenu plugins={[{label: 'entry_1', id: '1', integrationMode: 'iframe'}]}/>)
    expect(document.getElementsByClassName('sideMenu_icon')).toHaveLength(1)
  })

  it('show external link icon', () => {
    RenderWithReactIntl(
      <SideMenu
        plugins={[
          {label: 'entry_1', id: '1', integrationMode: 'href'},
          {label: 'entry_2', id: '2', integrationMode: 'iframe'}
        ]}
      />
    )
    expect(document.getElementsByClassName('sideMenu_icon')).toHaveLength(2)
    expect(document.getElementsByClassName('sideMenu_externalLink')).toHaveLength(1)
  })

  it('Generate menu using the configurations', () => {
    const plugins = [
      {
        id: 'plugin-test-2',
        label: 'Second test plugin',
        icon: 'home',
        order: 2,
        integrationMode: 'href',
        externalLink: {
          url: 'https://google.it',
          sameWindow: true
        }
      },
      {
        id: 'plugin-test-1',
        label: 'First test plugin',
        icon: 'clipboard',
        order: 1,
        integrationMode: 'href',
        externalLink: {
          url: 'https://google.it',
          sameWindow: false
        }
      },
      {
        id: 'plugin-test-3',
        label: 'IFrame',
        icon: 'clipboard',
        order: 1,
        integrationMode: 'iframe',
        pluginRoute: '/iframeTest',
        pluginUrl: 'https://www.google.com/webhp?igu=1'
      },
      {
        id: 'plugin-test-4',
        label: 'Qiankun',
        icon: 'clipboard',
        order: 1,
        integrationMode: 'qiankun',
        pluginRoute: '/qiankunTest',
        pluginUrl: 'https://www.google.com/webhp?igu=1'
      }
    ]
    RenderWithReactIntl(<SideMenu plugins={plugins}/>)

    expect(screen.getByText('First test plugin')).toBeTruthy()
    expect(screen.getByText('Second test plugin')).toBeTruthy()
    expect(screen.getByText('IFrame')).toBeTruthy()
    expect(screen.getByText('Qiankun')).toBeTruthy()
  })

  it('Overlay close menu correctly', () => {
    const setMenuOpened = jest.fn()
    const menuOpenedConfig = {
      isMenuOpened: true,
      setMenuOpened
    }
    RenderWithReactIntl(
      <MenuOpenedProvider value={menuOpenedConfig}>
        <SideMenu plugins={[]}/>
      </MenuOpenedProvider>
    )
    expect(setMenuOpened).not.toHaveBeenCalled()
    expect(document.getElementsByClassName('ant-drawer-open').length).toBe(1)
    expect(document.getElementsByClassName('ant-drawer-content').length).toBe(1)
    userEvent.click(document.getElementsByClassName('ant-drawer-mask')[0])
    expect(setMenuOpened).toHaveBeenCalledWith(false)
  })
})
