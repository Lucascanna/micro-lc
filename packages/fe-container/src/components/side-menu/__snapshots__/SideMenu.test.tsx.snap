// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideMenu tests match snapshot 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <div
      class="ant-drawer ant-drawer-left sideMenu_drawer"
      style="position: absolute;"
      tabindex="-1"
    >
      <div
        class="ant-drawer-mask"
      />
      <div
        class="ant-drawer-content-wrapper"
        style="transform: translateX(-100%); width: 378px;"
      >
        <div
          class="ant-drawer-content"
        >
          <div
            class="ant-drawer-wrapper-body"
          >
            <div
              class="ant-drawer-body"
            >
              <ul
                class="ant-menu ant-menu-root ant-menu-inline ant-menu-light sideBar"
                data-menu-list="true"
                role="menu"
                tabindex="0"
              >
                <li
                  class="ant-menu-item sideMenu_voice"
                  data-menu-id="rc-menu-uuid-test-1"
                  role="menuitem"
                  style="padding-left: 24px;"
                  tabindex="-1"
                >
                  <i
                    class="sideMenu_icon  ant-menu-item-icon"
                  />
                  <span
                    class="ant-menu-title-content"
                  >
                    <div
                      class="sideMenu_entry"
                    >
                      <span
                        class="sideMenu_label"
                      >
                        <span
                          class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
                        >
                          entry_1
                        </span>
                        <i
                          class="fas fa-external-link-alt sideMenu_externalLink"
                        />
                      </span>
                    </div>
                  </span>
                </li>
                <li
                  class="ant-menu-item sideMenu_voice"
                  data-menu-id="rc-menu-uuid-test-2"
                  role="menuitem"
                  style="padding-left: 24px;"
                  tabindex="-1"
                >
                  <i
                    class="sideMenu_icon  ant-menu-item-icon"
                  />
                  <span
                    class="ant-menu-title-content"
                  >
                    <div
                      class="sideMenu_entry"
                    >
                      <span
                        class="sideMenu_label"
                      >
                        <span
                          class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
                        >
                          entry_2
                        </span>
                        <i
                          class="fas fa-external-link-alt sideMenu_externalLink"
                        />
                      </span>
                    </div>
                  </span>
                </li>
              </ul>
              <div
                aria-hidden="true"
                style="display: none;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
